### DISCUSSION TITLE
🚀  GitHub Copilot Metrics API now available in public beta

### DISCUSSION THREAD

**Discussion by Akash1134** [community/community#120952] [Copilot News and Announcements]
We’re excited to announce the public beta of the GitHub Copilot Metrics API, available to all customers of **GitHub Copilot Business and GitHub Copilot Enterprise** starting today.

The [GitHub Copilot Metrics API](https://docs.github.com/en/rest/copilot/copilot-usage?apiVersion=2022-11-28) is designed to supply you with information about Copilot’s usage within your organizations. The data from the API is intended to be consumed and combined with your organization’s own data to create greater visibility into how Copilot fits into the bigger picture of your software development cycle.

To know what functionality will be included in the Public Beta, please refer the [changelog post](https://github.blog/changelog/2024-04-23-github-copilot-metrics-api-now-available-in-public-beta/#what-functionality-will-be-included-in-the-public-beta).

### Documentation and Resources
To help you get started, we have prepared comprehensive documentation and resources:

*   📝  API Documentation: Explore the detailed API documentation, including metrics definitions [here](https://docs.github.com/en/rest/copilot/copilot-usage?apiVersion=2022-11-28).
*   📊  Visualization: You can find an open source tool for consuming and visualizing data from the endpoint [here](https://github.com/github-copilot-community/copilot-metrics-viewer)
*   🧑‍🎓  Learning Pathway: You can find an extended article on measuring the impact of GitHub Copilot [here](https://resources.github.com/learn/pathways/copilot/essentials/measuring-the-impact-of-github-copilot/)

### Participate in the Public Beta!
Your feedback during this beta phase is invaluable to us. We encourage you to share your experiences, which will be instrumental in refining and enhancing the API as we look toward the future.

Stay tuned for updates and enhancements throughout the beta period. We’re committed to delivering a robust and feature-rich API that meets your needs and expectations.

**Comment by aline-viana** [comment]
Hi, we are very excited to test this resourse!
I made an api call using javascript to get information, but I receive the` status: 404` response. 
So, for some company this feature is not available yet? 
I try org and enterprise and the telemetry is active

**Reply by gmotodd** [reply]
Same case for me, calling the API through Postman and using fine-grained tokens on both, Orgs/{} and Orgs/{}/Teams{} URIs, return HTTP-404 as if our Org and Team(s) did not exist. 

I validated this by purposedly passing correct/incorrect URI elements ("org" instead of "orgs") to see whether the URI is valid, or my particular resources (Org and Team) do not exist, and I did get different responses.

When calling using "org" only, which is wrong, I get a 404 pointing me to the generic API documentation:
{
    "message": "Not Found",
    "documentation_url": "https://docs.github.com/rest"
}

And when sending the correct word "orgs", I get a 404 pointing me to this specific API's documentation:
{
    "message": "Not Found",
    "documentation_url": "https://docs.github.com/rest/copilot/copilot-usage#get-a-summary-of-copilot-usage-for-organization-members"
}

This tells me that our Org and Team(s) resources are just not found.

Another post in this forum below (https://github.com/orgs/community/discussions/112377) seems to indicate that the API worked for them.

Does anybody know if something else needs to be enabled Admin-wise in the github.com portal for the API to find our resources? We have been using GH Copilot for 6+ months now.

**Reply by SIkebe** [reply]
> calling the API through Postman and using fine-grained tokens

How about PAT (classic)? Fine-grained personal access tokens is not supported according to the docs.
> This endpoint does not work with GitHub App user access tokens, GitHub App installation access tokens, or fine-grained personal access tokens.
https://docs.github.com/en/rest/copilot/copilot-usage?apiVersion=2022-11-28#get-a-summary-of-copilot-usage-for-enterprise-members--fine-grained-access-tokens

**Reply by gmotodd** [reply]
Thanks for the thought. That is true for the Enterprises/{} URI. I am trying to consume the Orgs/{} and Orgs/{}/Teams{} URIs, for which the documentation does say that fine-grained personal access tokens are allowed.

![image](https://github.com/community/community/assets/122111056/033c35ac-08a4-4b3b-8b63-70c8feb4654c)

And if the token was an issue, shouldn't I receive either 401 or 403 errors instead of 404 - resource not found?

**Comment by erikvanlith** [comment]
Hi, I also have the same experience. 
* I tried with "Fine-grained personal access tokens". During the setup I added enough permissions and selected the organization. I get:
```
  "message": "Resource not accessible by personal access token",
  "documentation_url": "https://docs.github.com/rest/copilot/copilot-usage#get-a-summary-of-copilot-usage-for-organization-members"
```

* I also tried a classic token and then I get 
```
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest/copilot/copilot-usage#get-a-summary-of-copilot-usage-for-organization-members"
```

Are we missing something? We have a GitHub Copilot Business account.

**Comment by erikvanlith** [comment]
I think we found it and it is actually stated in the documentation (I have to admit ;) ): "_Copilot Business or Copilot Enterprise organization owners, and_ **owners** _and billing managers of their parent enterprises, can view Copilot usage metrics._". I personally have a github account and I am a member of the organization. 

With the organization account you can make someone member, but also owner. For this to work it seems you have to be an owner. We will try some things, but I hope this might already help some people.

**Reply by aline-viana** [reply]
Hi Erik! I try with the owner, but I have the response I posted

**Reply by erikvanlith** [reply]
Do you also use the "Fine-grained personal access token"?

**Reply by aline-viana** [reply]
yes

**Reply by gmotodd** [reply]
I also am an Enterprise and Org owner (not sure if you can get fine-grained tokens otherwise).
For kicks, I updated all my Account and Repository permissions to at least Read-Only, but to no avail still.

And again, if permissions are the issue, wouldn't the API responses be 401 (Unauthorized) or 403 (Forbidden), instead of 404 (Resource not found)?

Question: What is the difference between "PAT (classic)" tokens and fine-grained? And how do you obtain classic ones to give it a try?

**Reply by gmotodd** [reply]
Update: I tried calling the teams API (https://api.github.com/orgs/.../teams) with the same fine-grained token, and I did get an expected 403 saying that my token is not supported:

![image](https://github.com/community/community/assets/122111056/44295ccc-fc2f-43bb-8255-dbb92390455b)

But not getting the same response in the Copilot endpoint kind of validates that the token is not the problem, but rather the resource value I'm passing that it is not found (I triple-checked syntax), or something else need to be enabled inside GitHub.

**Reply by gmotodd** [reply]
And when calling the Orgs endpoint (https://api.github.com/orgs/{...}) with the same fine-grained token, it returns data.

I'm assuming this validates that the org resource value I'm passing is valid, just not for the Copilot metrics endpoint. Would this be a sign that something else needs to be enabled/configured in the GitHub portal admin-wise?

**Reply by gmotodd** [reply]
Solved! I tried passing a Personal Access Token (Classic) and it worked. Go figure!

**Reply by aline-viana** [reply]
@gmotodd I try with classic token and work for me
https://github.com/settings/personal-access-tokens/new click in Token (classic) and generate new

**Comment by ndemayo-bah** [comment]
Hello! Seeing something odd maybe in the metrics being reported... it looks like the `breakdown` portion of the reponse is returning duplicated results. First screenshot is the data converted to csv, second shows the raw response from the endpoint... is this something others have noticed?

![image](https://github.com/community/community/assets/114487562/a4c8f103-1b9d-41df-8c23-9bbf3115fc75)
![image](https://github.com/community/community/assets/114487562/22f4d69c-01b8-437c-ad96-496bdb65ae09)

**Reply by ndemayo-bah** [reply]
Looking a bit closner, it does appear as though it hasn't done this since 04/18, so maybe it's something that has been fixed already.

**Comment by DevOps-zhuang** [comment]
Team, for the copilot metric API, My customer found two issues, and I can reproduce it in my environment, we both use  Personal Access Token (Classic)  1) when calling organization level API like [https://api.github.com/orgs/{MyOrg}/copilot/usage](https://api.github.com/orgs/{org_name}/copilot/usage), it didn't return data on "day": "2024-04-05, the date before or after it can be returned, it is some interesting  2）the team level API(https://api.github.com/orgs/{org}/team/{team_slug}/copilot/usage) just returns empty, the http status code is 200, so it seems the authentication is right. and I am sure the team is over 20 people.  are those of known issues? thanks.

**Comment by Akash1134** [comment]
We’ve updated how we calculate `Last Activity` to give you better clarity and are pausing access to the Team endpoint in the Metrics API.

Ahead of the GitHub Copilot Metrics API launch, we made an update to how we calculate Last Activity in order to provide more useful information for admins. 

To know what has changed, please [read this changelog](https://github.blog/changelog/2024-05-09-github-copilot-metrics-updates/)

You can provide feedback and follow along for updates via this [Discussion](https://github.com/orgs/community/discussions/123032).

**Comment by sm-mbrt** [comment]
Hi! In our organization we have `x` seats purchased and Copilot enabled for selected `x` members.

I'm using Copilot Metrics API to gather usage data and for some days there is `total_active_users: x+1`.

My intuition tells me that this number should not be greater than the number of members with access to Copilot.

Is this something others have noticed?

**Comment by sm-mbrt** [comment]
Today the API does not return my statistics for the previous day, as it used to (I check it every day at the same time). I noticed that the days when no one used Copilot do not show up in the API's response, however, I am sure that yesterday people in my org were using it.

Update: a few days later statistic for that day are still missing

**Comment by sm-mbrt** [comment]
Today I don't have breakdown statistics for yesterday, `breakdown` is an empty array in the response although there are total values available:
```
{'breakdown': [],
  'day': '2024-06-06',
  'total_acceptances_count': ...,
  'total_active_chat_users': ...,
  'total_active_users': ...,
  'total_chat_acceptances': ...,
  'total_chat_turns': ...,
  'total_lines_accepted': ...,
  'total_lines_suggested': ...,
  'total_suggestions_count': ...}
```
All the issues I have encountered are related to [organization endpoint](https://docs.github.com/en/rest/copilot/copilot-usage?apiVersion=2022-11-28#get-a-summary-of-copilot-usage-for-organization-members).

### Update 14-06-2024:
Today, most of the total values for yesterday is 0 (acceptances count, suggestions count, lines suggested, lines accepted, active users), even though there are values for that metrics in a breakdown. Also, values for chat doesn't seem to be valid (too low).

At this point it's hard to rely on this API's responses.
Can I expect these topics to be addressed? Is this a widespread issue or just affecting me? I really care about this API because it helps me measure how Copilot impacts our developers' work and justify buying more seats.

### Update 15-06-2024:
Data for 14-06-2024 is missing.

### Update 20-06-2024:
Data for 19-06-2024 is missing. As you can see it's happening relatively often.

### Update 12.08-2024:
Lately, issues that I mentioned above, do not occur.

**Comment by WaterSwallow** [comment]
> We’re excited to announce the public beta of the GitHub Copilot Metrics API, available to all customers of **GitHub Copilot Business and GitHub Copilot Enterprise** starting today.
> 
> The [GitHub Copilot Metrics API](https://docs.github.com/en/rest/copilot/copilot-usage?apiVersion=2022-11-28) is designed to supply you with information about Copilot’s usage within your organizations. The data from the API is intended to be consumed and combined with your organization’s own data to create greater visibility into how Copilot fits into the bigger picture of your software development cycle.
> 
> To know what functionality will be included in the Public Beta, please refer the [changelog post](https://github.blog/changelog/2024-04-23-github-copilot-metrics-api-now-available-in-public-beta/#what-functionality-will-be-included-in-the-public-beta).
> 
> ### Documentation and Resources
> To help you get started, we have prepared comprehensive documentation and resources:
> 
> * 📝  API Documentation: Explore the detailed API documentation, including metrics definitions [here](https://docs.github.com/en/rest/copilot/copilot-usage?apiVersion=2022-11-28).
> * 📊  Visualization: You can find an open source tool for consuming and visualizing data from the endpoint [here](https://github.com/github-copilot-community/copilot-metrics-viewer)
> * 🧑‍🎓  Learning Pathway: You can find an extended article on measuring the impact of GitHub Copilot [here](https://resources.github.com/learn/pathways/copilot/essentials/measuring-the-impact-of-github-copilot/)
> 
> ### Participate in the Public Beta!
> Your feedback during this beta phase is invaluable to us. We encourage you to share your experiences, which will be instrumental in refining and enhancing the API as we look toward the future.
> 
> Stay tuned for updates and enhancements throughout the beta period. We’re committed to delivering a robust and feature-rich API that meets your needs and expectations.

Exciting news - GitHub has launched a public beta of the GitHub Copilot Metrics API. This API will provide valuable information about Copilot’s usage within your organizations which can be combined with your own data for insights into your software development cycle. Detailed information about what's included in the beta can be found in the supplied changelog post, API documentation, and resources for data visualization and learning. Feedback during this beta phase is greatly appreciated as it would guide the refinement of the API.

**Comment by GeorgeChiran** [comment]
Can someone point me to some documentation on how to use the team level aggregates? I"m using the API looks good, but will be good to have something more granular.
